File: priv/def/test/exdl_svg.erl
%%
%%  Copyright (c) 2012 David Koch
%%
%%  See the file "license.terms" for information on usage and redistribution
%%  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
%%
%%     $Id: exdl_svg.erl,v 1.5 2005/06/29 14:45:38 dgud Exp $
%%

-module(exdl_svg).

-include("exdl.hrl").

-ifdef(COMPILE_SVG).

%% 1- EXPORT

%-compile(export_all).

-export( [ load/1
         , loadBuffer/2
         , setOffset/3
         , setScale/3
         , renderToSurface/4
         , free/1
         , setFlags/2
         , width/1
         , height/1
         , version/0
         , setClipping/5
         ]
       ).

%% 2- INCLUDE

-include("exdl_svg.hrl").

%% 3- IMPORT

-import( exdl_driver
       , [ call/2
         , cast/2
         ]
       ).

%% 4- FUNCTIONS

%% Func:       load
%% Args:       FileName
%% Returns:    Exdl
%% C-API func: SDL_svg_context* SVG_Load(const char* FileName)

load(FileName)
 -> case
      call
      ( ?SVG_Load
      , <<FileName:?_PTR
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       loadBuffer
%% Args:       Data, Len
%% Returns:    Exdl
%% C-API func: SDL_svg_context* SVG_LoadBuffer(char* data, int len)

loadBuffer(Data, Len)
 -> case
      call
      ( ?SVG_LoadBuffer
      , <<Data:?_PTR
        , Len:?Int
        >>
      )
    of
      0
        -> { nullptr, ?MODULE, ?LINE }
    ; <<Exdl:?_PTR
      >>
        -> { buffptr, Exdl }
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       setOffset
%% Args:       Source, AOffsetX, AOffsetY
%% Returns:    Exdl
%% C-API func: int SVG_SetOffset(SDL_svg_context* Source, double aOffsetX, double aOffsetY)

setOffset(Source, AOffsetX, AOffsetY)
 -> case
      call
      ( ?SVG_SetOffset
      , <<Source:?_PTR
        , AOffsetX:?Double
        , AOffsetY:?Double
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       setScale
%% Args:       Source, AScaleX, AScaleY
%% Returns:    Exdl
%% C-API func: int SVG_SetScale(SDL_svg_context* Source, double aScaleX, double aScaleY)

setScale(Source, AScaleX, AScaleY)
 -> case
      call
      ( ?SVG_SetScale
      , <<Source:?_PTR
        , AScaleX:?Double
        , AScaleY:?Double
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       renderToSurface
%% Args:       Source, X, Y, Target
%% Returns:    Exdl
%% C-API func: int SVG_RenderToSurface(SDL_svg_context* Source, int X, int Y, SDL_Surface* Target)

renderToSurface(Source, X, Y, Target)
 -> case
      call
      ( ?SVG_RenderToSurface
      , <<Source:?_PTR
        , X:?Int
        , Y:?Int
        , Target:?_PTR
        >>
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       free
%% Args:       Source
%% Returns:    ok
%% C-API func: void SVG_Free(SDL_svg_context* Source)

free(Source)
 -> cast
    ( ?SVG_Free
    , <<Source:?_PTR
      >>
    )
.

%% Func:       setFlags
%% Args:       C, Flags
%% Returns:    ok
%% C-API func: void SVG_Set_Flags(SDL_svg_context* c, unsigned long flags)

setFlags(C, Flags)
 -> cast
    ( ?SVG_Set_Flags
    , <<C:?_PTR
      , Flags:?Ulong
      >>
    )
.

%% Func:       width
%% Args:       C
%% Returns:    Exdl
%% C-API func: float SVG_Width(SDL_svg_context* c)

width(C)
 -> case
      call
      ( ?SVG_Width
      , <<C:?_PTR
        >>
      )
    of
      <<Exdl:?Float
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       height
%% Args:       C
%% Returns:    Exdl
%% C-API func: float SVG_Height(SDL_svg_context* c)

height(C)
 -> case
      call
      ( ?SVG_Height
      , <<C:?_PTR
        >>
      )
    of
      <<Exdl:?Float
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       version
%% Args:       none
%% Returns:    Exdl
%% C-API func: int SVG_Version(void)

version()
 -> case
      call
      ( ?SVG_Version
      , [
        ]
      )
    of
      <<Exdl:?Int
      >>
        -> Exdl
    ; Res
        -> { badarg, ?MODULE, ?LINE, Res }
    end
.

%% Func:       setClipping
%% Args:       C, Minx, Miny, Maxx, Maxy
%% Returns:    ok
%% C-API func: void SVG_SetClipping(SDL_svg_context* c, int minx, int miny, int maxx, int maxy)

setClipping(C, Minx, Miny, Maxx, Maxy)
 -> cast
    ( ?SVG_SetClipping
    , <<C:?_PTR
      , Minx:?Int
      , Miny:?Int
      , Maxx:?Int
      , Maxy:?Int
      >>
    )
.


-endif. % COMPILE_SVG

File: priv/def/test/exdl_svg.c
/*
 *  Copyright (c) 2012 David Koch
 *
 *  See the file "license.terms" for information on usage and redistribution
 *  of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 *     $Id: exdl_svg.c,v 1.5 2005/06/29 14:45:38 dgud Exp $
 */

#include "exdl.h"

#ifdef COMPILE_SVG

EXDL_API(esvg_load)
{
  /* SDL_svg_context* SVG_Load(const char* FileName) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(char, FileName[])
  STACK_FMT_END()

  CATCH_RESULT(SDL_svg_context*)

    SVG_Load
    ( STACK_GET_ADDR(FileName)
    );

  THROW_POINTER()
}

EXDL_API(esvg_loadBuffer)
{
  /* SDL_svg_context* SVG_LoadBuffer(char* data, int len) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(char*, data)
    STACK_FMT_DATA(int,   len)
  STACK_FMT_END()

  CATCH_RESULT(SDL_svg_context*)

    SVG_LoadBuffer
    ( STACK_GET_DATA(data)
    , STACK_GET_DATA(len)
    );

  THROW_POINTER()
}

EXDL_API(esvg_setOffset)
{
  /* int SVG_SetOffset(SDL_svg_context* Source, double aOffsetX, double aOffsetY) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_svg_context*, Source)
    STACK_FMT_DATA(double,           aOffsetX)
    STACK_FMT_DATA(double,           aOffsetY)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SVG_SetOffset
    ( STACK_GET_DATA(Source)
    , STACK_GET_DATA(aOffsetX)
    , STACK_GET_DATA(aOffsetY)
    );

  THROW_RESULT(int)
}

EXDL_API(esvg_setScale)
{
  /* int SVG_SetScale(SDL_svg_context* Source, double aScaleX, double aScaleY) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_svg_context*, Source)
    STACK_FMT_DATA(double,           aScaleX)
    STACK_FMT_DATA(double,           aScaleY)
  STACK_FMT_END()

  CATCH_RESULT(int)

    SVG_SetScale
    ( STACK_GET_DATA(Source)
    , STACK_GET_DATA(aScaleX)
    , STACK_GET_DATA(aScaleY)
    );

  THROW_RESULT(int)
}

EXDL_API(esvg_renderToSurface)
{
  /* int SVG_RenderToSurface(SDL_svg_context* Source, int X, int Y, SDL_Surface* Target) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_svg_context*, Source)
    STACK_FMT_DATA(int,              X)
    STACK_FMT_DATA(int,              Y)
    STACK_FMT_DATA(SDL_Surface,      Target[])
  STACK_FMT_END()

  CATCH_RESULT(int)

    SVG_RenderToSurface
    ( STACK_GET_DATA(Source)
    , STACK_GET_DATA(X)
    , STACK_GET_DATA(Y)
    , STACK_GET_ADDR(Target)
    );

  THROW_RESULT(int)
}

EXDL_API(esvg_free)
{
  /* void SVG_Free(SDL_svg_context* Source) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_svg_context, Source[])
  STACK_FMT_END()

    SVG_Free
    ( STACK_GET_ADDR(Source)
    );
}

EXDL_API(esvg_set_Flags)
{
  /* void SVG_Set_Flags(SDL_svg_context* c, unsigned long flags) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_svg_context*, c)
    STACK_FMT_DATA(unsigned long,    flags)
  STACK_FMT_END()

    SVG_Set_Flags
    ( STACK_GET_DATA(c)
    , STACK_GET_DATA(flags)
    );
}

EXDL_API(esvg_width)
{
  /* float SVG_Width(SDL_svg_context* c) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_svg_context, c[])
  STACK_FMT_END()

  CATCH_RESULT(float)

    SVG_Width
    ( STACK_GET_ADDR(c)
    );

  THROW_RESULT(float)
}

EXDL_API(esvg_height)
{
  /* float SVG_Height(SDL_svg_context* c) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_DATA(SDL_svg_context, c[])
  STACK_FMT_END()

  CATCH_RESULT(float)

    SVG_Height
    ( STACK_GET_ADDR(c)
    );

  THROW_RESULT(float)
}

EXDL_API(esvg_version)
{
  /* int SVG_Version(void) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  CATCH_RESULT(int)

    SVG_Version
    (
    );

  THROW_RESULT(int)
}

EXDL_API(esvg_setClipping)
{
  /* void SVG_SetClipping(SDL_svg_context* c, int minx, int miny, int maxx, int maxy) */

  /* exp :  */
  /* inp :  */
  /* ret :  */

  STACK_FMT_BEGIN()
    STACK_FMT_LPTR(SDL_svg_context*, c)
    STACK_FMT_DATA(int,              minx)
    STACK_FMT_DATA(int,              miny)
    STACK_FMT_DATA(int,              maxx)
    STACK_FMT_DATA(int,              maxy)
  STACK_FMT_END()

    SVG_SetClipping
    ( STACK_GET_DATA(c)
    , STACK_GET_DATA(minx)
    , STACK_GET_DATA(miny)
    , STACK_GET_DATA(maxx)
    , STACK_GET_DATA(maxy)
    );
}

#endif // COMPILE_SVG
